### DISCUSSION TITLE
Cancel other jobs in GitHub Actions workflow when one job fails (fail-fast across workflow)

### DISCUSSION THREAD

**Discussion by Piedone** [community/community#38361] [Actions]
When executing the same job on multiple runners in a matrix, we have the [`fail-fast`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast) option that cancels all instances of the job when one fails. Is there something similar that works across jobs in a workflow? I.e. if you have multiple jobs in a workflow, and one fails, cancel the other ones?

This only seems to be possible by interacting with the [GitHub web API](https://docs.github.com/en/rest/actions/workflow-runs#cancel-a-workflow-run). Is there something simpler that I'm missing?

Thank you!

**Comment by david-engelmann** [comment]
If the jobs are order dependent then you can have them depend on each other. job 2 won't go without a successful job 1. Check out the snippet below to get an idea.

```
jobs:
  job1:
  job2:
    needs: job1
  job3:
    needs: [job1, job2]
```

**Reply by Piedone** [reply]
I didn't specify, but the jobs don't depend on each other (otherwise this wouldn't be a challenge as you also point out). We have three jobs, and while we don't want to lose out on parallelization, we don't want to waste resources on the rest if one of them fails (this being a build workflow, any of the jobs failing means the whole build is in a failed state). Thank you though!

**Comment by Piedone** [comment]
While I still don't think there's a built-in way to do this, apart from doing the API call directly (which is simple enough), you can use the [Cancel Workflow Action for failed jobs](https://github.com/marketplace/actions/cancel-workflow-action-for-failed-jobs) or [Cancel this build](https://github.com/marketplace/actions/cancel-this-build) actions. However, I don't think either of them are actively maintained anymore.

**Reply by dsherret** [reply]
The GitHub API seems to only work when using a PAT. I tried the following at the end of a workflow:

```yml
- run: exit 1 # for testing
- name: Cancel build on failure
   if: failure()
   run: |-
     curl \
       -X POST \
       -H "X-GitHub-Api-Version: 2022-11-28" \
       -H "Accept: application/vnd.github.v3+json" \
       -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
       "https://api.github.com/repos/$GITHUB_REPOSITORY/actions/runs/${{ github.run_id }}/cancel"
```

And the response I get back is:

```json
{
  "message": "Resource not accessible by integration",
  "documentation_url": "https://docs.github.com/rest/reference/actions#cancel-a-workflow-run"
}
```

**Reply by Piedone** [reply]
Interesting, because that worked for us: https://github.com/Lombiq/GitHub-Actions/blob/635bad77e0144518be5475a8de98aeebcd13bb9a/.github/actions/cancel-workflow/action.yml We since moved this to use the GitHub CLI, but that still authenticates with `GITHUB_TOKEN`: https://github.com/Lombiq/GitHub-Actions/blob/dev/.github/actions/cancel-workflow/action.yml

**Reply by dsherret** [reply]
Hmmm... maybe I was doing something slightly wrong. Oh well, I've given up on this already. There should be an easier built-in way to do this.

**Reply by jsoref** [reply]
You might have a permission restricted organization. Try explicitly adding:

```yaml
permissions:
  actions: write
```

**Reply by fearocanity** [reply]
Using the workflow token (GITHUB API)  works fine to me, it didn't appear that error. Using @dsherret snippet

**Reply by jsoref** [reply]
Fwiw, odds are that @dsherret's orgs were set to the [restrictive setting for `GITHUB_TOKEN`](https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization#configuring-the-default-github_token-permissions) whereas the people who didn't need to add permissions were set to the overly permissive setting -- mostly the default would depend on when an account/org was created or whether the owner has toggled the value (my original account is old enough that it had the permissive default which I've changed to the restrictive one, and I have new orgs that were created w/ the restrictive default).

**Comment by segevmalool** [comment]
A conditional on this action works pretty well! https://github.com/marketplace/actions/cancel-this-build

**Reply by pjgrizel** [reply]
It cancels the build (very well), but doesn't mark it as 'failed' which can be confusing

**Comment by alyssaruth** [comment]
I would also really like this feature - we have a pipeline that runs a lot of things in parallel - some jobs have dependencies, others do not. Here's a sub-set to illustrate:

```
Frontend tests
Frontend build -> Publish docker image -------
Backend tests                                |
Backend build -> Publish docker image -> Deploy -> Run Integration Tests (matrix)
```

If the frontend or backend tests fail, I would like to abort the whole pipeline to save on resource and give faster feedback to engineers. However, there is no dependency on them - I don't want "Deploy" to wait for all tests to finish, because it doesn't need any outputs from them.

We've recently moved over from Jenkins, which had a global `parallelsAlwaysFailFast()` setting. Having an equivalent global option for an actions pipeline would be awesome!

**Comment by alyssaruth** [comment]
In case it's useful to anyone, we've ended up working around this using a post-job script in our self-hosted runners. More about those here: https://docs.github.com/en/actions/hosting-your-own-runners/running-scripts-before-or-after-a-job

The crux of the python code is this part:

```
def process_workflow_for_fail_fast(github_run_id: int, repo_id: int):
    client, token = get_github_oauth_client()
    workflow = client.get_repo(repo_id).get_workflow_run(github_run_id)
    head = {"Authorization": "Bearer {}".format(token)}
    jobs = requests.get(url=workflow.jobs_url, headers=head).json()
    statuses = (job["conclusion"] for job in jobs["jobs"])
    if "failure" in statuses:
        print("Failure found. Cancelling workflow!")
        workflow.cancel()
```

We'd still prefer not to have to maintain this if we could avoid it - it's a bit sneaky/hidden for my liking. It will also only work if you're self-hosting your runners. But it does the job for us!

**Reply by alexef** [reply]
thanks for this

**Comment by pedro-beemon** [comment]
I found this solution [https://github.com/andymckay/cancel-action](https://github.com/community/community/discussions/url).
Hope it makes easyer our lifes.

**Reply by knocte** [reply]
broken link

**Reply by jsoref** [reply]
That's strange, the link text is fine: https://github.com/andymckay/cancel-action
(The link itself is just broken)

**Comment by ivankovnatsky** [comment]
We've trying it like this:

```
      - if: "failure()"
        uses: "andymckay/cancel-action@0.3"
```

We'll see how it goes.

**Reply by mtvx** [reply]
How did it go?

**Reply by ivankovnatsky** [reply]
Our pipelines are pretty extensive, long rust builds, tests, and judging by initial evaluation it does the job pretty good. Meaning we don't have to wait for other jobs, when some of the job's fail will just trigger workflow cancel, marvelous. Time saver.

Only, we had to add this permission: https://github.com/andymckay/cancel-action/issues/4#issuecomment-1618730540.

**Reply by ivankovnatsky** [reply]
Currently it's working the way we expected, really helping to avoid long pipeline runs of unnecessary jobs.

**Comment by droznyk** [comment]
For people that will land here searching for the solution:
GitHub provides a "fail-fast" feature in their GitHub Workflows: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

**Reply by jsoref** [reply]
No.

This is not asking about matrixes. It's asking for a way to treat manually declared jobs like matrix jobs and provide a fail-fast for the manually declared parallel jobs.
